### PR TITLE
ModNioResourcePack: don't extend AbstractFileResourcePack

### PR DESCRIPTION
**Pull request by JFronny**
When a mod comes across an AbstractFileResourcePack, it should be safe to assume that the pack is a resource pack based on a File (as the name suggests).
That is not true in this implementation as the attached "base" File is null.
This PR adjusts ModNioResourcePack to no longer extend AbstractFileResourcePack and instead implement the small number of features it provided manually.

### CODE REVIEW

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java [hunk 1]
```diff
@@ -51,9 +53,9 @@
 import net.fabricmc.loader.api.ModContainer;
 import net.fabricmc.loader.api.metadata.ModMetadata;
 
-public class ModNioResourcePack extends AbstractFileResourcePack implements ModResourcePack {
+public class ModNioResourcePack implements ResourcePack, ModResourcePack {
 	private static final Logger LOGGER = LoggerFactory.getLogger(ModNioResourcePack.class);
-	private static final Pattern RESOURCE_PACK_PATH = Pattern.compile("[a-z0-9-_.]+");
+	private static final Pattern RESOURCE_PACK_PATH = Pattern.compile("[a-z\\d-_.]+");
```
Comment (apple502j) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: Why this change? Won't this allow non-ASCII paths? (like Persian numerals)

Comment (JFronny) [reply fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: Idea said it was equivalent

Comment (Juuxel) [reply fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: ...why would you change it then? 🤔

Comment (JFronny) [reply fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: You mean why I reverted that change?
It's not really related to this PR and idea might just be recommending the change for international compatibility
The reason why I changed it in the first place was that it was recommended by Idea.

Comment (Juuxel) [reply fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: I was wondering more about why you changed it originally since the ascii pattern is perfectly clear, but yeah it makes sense to revert it here

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java [hunk 2]
```diff
@@ -51,7 +53,7 @@
 import net.fabricmc.loader.api.ModContainer;
 import net.fabricmc.loader.api.metadata.ModMetadata;
 
-public class ModNioResourcePack extends AbstractFileResourcePack implements ModResourcePack {
+public class ModNioResourcePack implements ResourcePack, ModResourcePack {
```
Comment (apple502j) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java:59]: Since we're editing this file, could you put an `@ApiStatus.Internal` here?

Comment (JFronny) [reply fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java:59]: I think the fact that it is in the impl package implies that, but sure

**Review by apple502j — APPROVED** [review]

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java [hunk 3]
```diff
@@ -183,7 +185,6 @@ private boolean hasAbsentNs(String filename) {
 		return !namespaces.get(type).contains(filename.substring(prefixLen, nsEnd));
 	}
 
-	@Override
 	protected InputStream openFile(String filename) throws IOException {
```
Comment (modmuss50) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: Can this be private now?

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java [hunk 4]
```diff
@@ -204,13 +205,17 @@ protected InputStream openFile(String filename) throws IOException {
...
 	@Override
-	protected boolean containsFile(String filename) {
-		if (ModResourcePackUtil.containsDefault(modInfo, filename)) {
-			return true;
+	public InputStream openRoot(String fileName) throws IOException {
+		if (!fileName.contains("/") && !fileName.contains("\\")) {
+			return this.openFile(fileName);
+		} else {
+			throw new IllegalArgumentException("Root resources can only be filenames, not paths (no / allowed!)");
 		}
```
Comment (modmuss50) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: [Nit] Would be slightly cleaner/easier to read to do `if(fileName.contains("/") || fileName.contains("\\")` and then throw. Remove the need for the else.

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java [hunk 5]
```diff
@@ -252,11 +257,30 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
...
+	}
+
 	@Override
 	public Set<String> getNamespaces(ResourceType type) {
 		return namespaces.getOrDefault(type, Collections.emptySet());
 	}
 
+	@Override
+	public <T> T parseMetadata(ResourceMetadataReader<T> metaReader) throws IOException {
+		try (InputStream is = openRoot("pack.mcmeta")) {
```
Comment (modmuss50) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: Can use `openFile`, no need to validate the path here as we know its fine.

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java [hunk 6]
```diff
@@ -290,4 +314,8 @@ private static boolean exists(Path path) {
 		// NIO Files.exists is notoriously slow when checking the file system
 		return path.getFileSystem() == DEFAULT_FS ? path.toFile().exists() : Files.exists(path);
 	}
+
+	private static String getFilename(ResourceType type, Identifier id) {
+		return type.getDirectory() + "/" + id.getNamespace() + "/" + id.getPath();
+	}
```
Comment (modmuss50) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: Use String.format same as vanilla, or even an access widener to the vanilla method might be nicer.

**Review by modmuss50 — CHANGES_REQUESTED** [review]

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java (hunk 6 above)
Comment (JFronny) [reply fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: I use the same String.format used in vanilla now

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java (hunk 5 above)
Comment (JFronny) [reply fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: OK

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java (hunk 3 above)
Comment (JFronny) [reply fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java]: Seems to be the case

**Review by modmuss50 — APPROVED** [review]

**Review by Technici4n — APPROVED** [review]
Had a look at `AbstractFileResourcePack` and this seems fine.

**Comment by modmuss50** [comment]
No need for last call as its not an API change, I will test with a handful of mods before merging.
